### PR TITLE
Concurrency

### PR DESCRIPTION
**Pull request by runnersaw**
Moves all long-running tasks to background thread and uses callbacks for when they're done.

### CODE REVIEW

**Comment by runnersaw** [comment]
The general structure is: handle request -> spin off thread -> return from http request -> callback gets called by thread

File: app/handlers/safetyAnalysis.py
Comment (dcelik) [inline app/handlers/safetyAnalysis.py]: Doesn't this need to be "return self.callback(...)"? You do the return in objectTracking but not here. If they are interchangeable, lets stay consistent and pick one.

File: app/handlers/safetyAnalysis.py
Comment (dcelik) [inline app/handlers/safetyAnalysis.py]: Same thing here.

File: app/handlers/testConfig.py
Comment (dcelik) [inline app/handlers/testConfig.py]: Same, be consistent within the codebase, especially within the same class.

File: app/handlers/testConfig.py
Comment (dcelik) [inline app/handlers/testConfig.py]: See comment above.

File: app/handlers/testConfig.py
Comment (dcelik) [inline app/handlers/testConfig.py]: Could it be possible to merge all these into a single helper method that takes in a video filename?

File: app/handlers/testConfig.py
Comment (dcelik) [inline app/handlers/testConfig.py]: We need another check here incase they pass in an invalid flag, right now it would just return success.

File: app/handlers/testConfig.py
Comment (dcelik) [inline app/handlers/testConfig.py]: Remove this print statement?

File: app/handlers/testConfig.py
Comment (dcelik) [inline app/handlers/testConfig.py]: Unfortunately, try:except with subprocess.call will hide exceptions and return success. In this case we should instead use subprocess.check_output as this will raise errors that happen within the subprocess calls. @jabb1123 can fill you in on why we need this.

File: app/handlers/testConfig.py
Comment (dcelik) [inline app/handlers/testConfig.py]: This is duplicated code, is it possible to make this a separate helper method?

File: app/traffic_cloud_utils/video.py
Comment (dcelik) [inline app/traffic_cloud_utils/video.py:17]: :(

**Review by dcelik — CHANGES_REQUESTED** [review]
High level feedback:
Everything looks pretty good, some issues with consistency of "return self.callback(..)" vs. "self.callback(...)", some code that could possibly be modularized, and the tons of subprocess.call's that I believe should be changed to subprocess.check_output. Those issues should be addressed before approval.

File: app/handlers/safetyAnalysis.py
Comment (runnersaw) [reply app/handlers/safetyAnalysis.py]: The `return` in the other methods is intended to stop execution of the function. In other functions, we don't want to keep doing work if something failed. Here, it doesn't matter. For consistency, this should be changed to `return self.callback()`.

BTW, it isn't actually important that it's returning the result of `self.callback()` because `self.callback()` returns nothing. It's simply meant to end execution of the function, and saves a line compared to:
```
self.callback()
return
```

File: app/handlers/testConfig.py
Comment (runnersaw) [reply app/handlers/testConfig.py]: Sure!

File: app/handlers/testConfig.py
Comment (runnersaw) [reply app/handlers/testConfig.py]: Good catch!

File: app/handlers/safetyAnalysis.py
Comment (dcelik) [reply app/handlers/safetyAnalysis.py]: Okay sounds good.

File: app/handlers/testConfig.py
Comment (jabb1123) [inline app/handlers/testConfig.py]: This change was already pushed to `dev`. I recommend pulling from `dev` to prevent conflicts or overwriting approved code.

File: app/handlers/safetyAnalysis.py
Comment (jabb1123) [inline app/handlers/safetyAnalysis.py]: I know that tyhe focus here wasn't the `subprocess.call()` but this needs to be `subprocess.check_output()` because we want the process to return the error that occurred and through an exception to prevent the rest of the function from executing. This change is currently on `dev` and this PR should be rejected until the changes here reflect that.

Comment (jabb1123) [reply app/handlers/safetyAnalysis.py]: I know that the focus here wasn't the `subprocess.call()` but this needs to be `subprocess.check_output()` because we want the process to return the error that occurred and through an exception to prevent the rest of the function from executing. This change is currently on `dev` and this PR should be rejected until the changes here reflect that.

File: app/handlers/testConfig.py
Comment (jabb1123) [reply app/handlers/testConfig.py]: This change was already pushed to `dev`. I recommend pulling from `dev` to prevent conflicts or overwriting approved code.

**Review by jabb1123 — CHANGES_REQUESTED** [review]

**Comment by jabb1123** [comment]
All conflicts resolved. Needs testing. Pull and test BEFORE MERGE!

**Review by jabb1123 — COMMENTED** [review]
Changes look solid now. Will approve after testing

**Comment by runnersaw** [comment]
As soon as create highlight video fixes get on dev (Deniz is working on them), I will fix that file also and this will be ready for final PR

**Comment by runnersaw** [comment]
Ready for final PR. @dcelik @jabb1123 go ahead and re-review.

File: app/traffic_cloud_utils/emailHelper.py
Comment (segerphilip) [inline app/traffic_cloud_utils/emailHelper.py:17]: :smile: :+1:

File: app/traffic_cloud_utils/app_config.py
Comment (jabb1123) [inline app/traffic_cloud_utils/app_config.py:138]: 🤦‍♂️

File: app/traffic_cloud_utils/video.py
Comment (jabb1123) [inline app/traffic_cloud_utils/video.py]: https://goo.gl/KCcV2d

File: app/traffic_cloud_utils/video.py
Comment (jabb1123) [inline app/traffic_cloud_utils/video.py]: 💀

File: app/traffic_cloud_utils/video.py
Comment (jabb1123) [inline app/traffic_cloud_utils/video.py]: 🛑

File: app/traffic_cloud_utils/pm.py
Comment (jabb1123) [inline app/traffic_cloud_utils/pm.py]: 🙅‍♂️

File: app/traffic_cloud_utils/pm.py
Comment (jabb1123) [inline app/traffic_cloud_utils/pm.py]: 🥇

File: app/traffic_cloud_utils/video.py
Comment (jabb1123) [inline app/traffic_cloud_utils/video.py]: https://goo.gl/bL40EV

**Review by jabb1123 — APPROVED** [review]
:)

File: app/traffic_cloud_utils/pm.py
Comment (dcelik) [inline app/traffic_cloud_utils/pm.py:36]: :(

**Review by dcelik — APPROVED** [review]
Looks good! The fixes for highlightVideo have been postponed due to graphics issues on vagrant. @youralien should have better luck then I at fixing them. @jabb1123 please don't pollute PR history with unnecessary comments... they have been removed.

**Pull request merged**
